store/smt/store.go

```diff
@@ -0,0 +1,211 @@
+package smt
+
+import (
+	"crypto/sha256"
+	"encoding/binary"
+	"io"
+	"sync"
+	"time"
+
+	dbm "github.com/cosmos/cosmos-sdk/db"
+	"github.com/cosmos/cosmos-sdk/store/cachekv"
+	"github.com/cosmos/cosmos-sdk/store/tracekv"
+	"github.com/cosmos/cosmos-sdk/store/types"
+	"github.com/cosmos/cosmos-sdk/telemetry"
+	abci "github.com/tendermint/tendermint/abci/types"
+
+	"github.com/lazyledger/smt"
+)
+
+var (
+	_ types.KVStore                 = (*Store)(nil)
+	_ types.CommitStore             = (*Store)(nil)
+	_ types.CommitKVStore           = (*Store)(nil)
+	_ types.Queryable               = (*Store)(nil)
+	_ types.StoreWithInitialVersion = (*Store)(nil)
+)
+
+var (
+	prefixLen      = 1
+	versionsPrefix = []byte{0}
+	dataPrefix     = []byte{1}
+	indexPrefix    = []byte{2}
+	afterIndex     = []byte{3}
+)
+
+// Store Implements types.KVStore and CommitKVStore.
+type Store struct {
+	tree *smt.SparseMerkleTree
+	db   dbm.DBReadWriter
+
+	version int64
+
+	opts struct {
+		initialVersion int64
+		pruningOptions types.PruningOptions
+	}
+
+	mtx sync.RWMutex
+}
+
+func NewStore(underlyingDB dbm.DBReadWriter) *Store {
+	return &Store{
+		tree: smt.NewSparseMerkleTree(underlyingDB, sha256.New()),
+		db:   underlyingDB,
+	}
+}
+
+// KVStore interface below:
+
+func (s *Store) GetStoreType() types.StoreType {
+	return types.StoreTypeSMT
+}
+
+// CacheWrap branches a store.
+func (s *Store) CacheWrap() types.CacheWrap {
+	return cachekv.NewStore(s)
+}
+
+// CacheWrapWithTrace branches a store with tracing enabled.
+func (s *Store) CacheWrapWithTrace(w io.Writer, tc types.TraceContext) types.CacheWrap {
+	return cachekv.NewStore(tracekv.NewStore(s, w, tc))
+}
+
+// Get returns nil iff key doesn't exist. Panics on nil key.
+func (s *Store) Get(key []byte) []byte {
+	defer telemetry.MeasureSince(time.Now(), "store", "smt", "get")
+	val, err := s.db.Get(dataKey(key))
+	if err != nil {
+		panic(err)
```

Comment:
we want to go into returning errors by default, rather than panics

Reply:
Panic is the only option here, since it's part of the `KVStore` interface.

Reply:
I know, but if we are updating it, then maybe we can "fix" the interface. @aaronc  - do you think this kind of change will be acceptable finally?